### PR TITLE
Fix java attach setting OBI uid/gid as 0s

### LINKED ISSUE #3018: JavaInjector.NewExecutable calls Cleanup with myUID=0 for non-Java processes, elevating non-root agents to euid 0
### Component

OBI (eBPF instrumentation)

### What happened?

When OBI runs as a **non-root** process with ambient `CAP_SETUID`/`CAP_SETGID` and `javaagent.enabled=true`, instrumenting **any non-Java** process (C++, Python, etc.) permanently elevates the OBI agent's effective UID to `0`.
This happens because `JavaInjector.NewExecutable` is invoked for every instrumentable, creates a `JAttacher` whose `myUID`/`myGID` are still the Go zero value (`0`), defers `Cleanup()`, and only then gates the Java attach path. For non-Java processes the attach path is skipped, but the deferred cleanup still runs and executes `syscall.Seteuid(0)` / `syscall.Setegid(0)`.

### What did you expect to happen?

Instrumenting a non-Java process must not change the OBI process credentials. Cleanup() should either not run, or restore the UID/GID captured by Init().

### Steps to reproduce

1. Run OBI (or an embedding agent that starts `instrumenter.Run` with a dynamic PID selector) as non-root with `CAP_SETUID`/`CAP_SETGID` and `Java.Enabled=true`.
2. Confirm agent identity before instrumentation:
   ```bash
   ps -o user=,pid=,cmd= -C <obi-or-agent-binary>
   # expect non-root, e.g. linux-user

### OBI configuration

```yaml

```

### Relevant logs

```shell

```

### Additional context

_No response_

### Version

_No response_

### Kernel version

_No response_

### OS / Distribution

_No response_

### Deployment

Bare metal / VM

### PR DESCRIPTION
**Pull request by grcevski**
## Summary

OBI java attach could set the uid/gid on the OBI process itself to 0 is undesirable. This PR fixes the flow with the following approach:

- We now pass the cancellation context to the attach helpers, so they would get cancelled if the timeout expired.
- We make the init() and cleanup() idempotent and only run if there was successful save of credentials. This makes the outer cleanup on timeout termination safe to be called when no init happened.
- We skip the whole logic if the process is not Java.

Fixes https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/issues/3018

## Validation

- [X] I have read and followed the [contributing guidelines](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/main/CONTRIBUTING.md)
- [ ] If this enhances / fixes / changes a core feature, I have updated the [features documentation](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/main/devdocs/features.md) and [support matrix](https://github.com/open-telemetry/opentelemetry-ebpf-instrumentation/blob/main/SUPPORT_MATRIX.md) as needed.

### CODE REVIEW

**Comment by grcevski** [comment]
I also tested manually with the following approach:

1. I set the capabilities on a freshly built OBI like this:
```
sudo setcap 'cap_sys_admin,cap_sys_ptrace,cap_dac_read_search,cap_net_raw,cap_perfmon,cap_bpf,cap_checkpoint_restore=ep' bin/obi
```
2. I ran a Java application and I stopped it with GDB, so the JVM will never respond.
3. I ran OBI to instrument the application. After the 10 seconds timeout I saw:
```
time=2026-08-13T14:31:15.649-04:00 level=WARN msg="java attach timed out" component=javaagent.Injector timeout=10s pid=235034
time=2026-08-13T14:31:15.650-04:00 level=WARN msg="unable to attach java agent to process, Java TLS telemetry will not work" component=discover.traceAttacher pid=235034 error="java attach timed out"
time=2026-08-13T14:31:15.650-04:00 level=INFO msg="instrumenting process" component=discover.traceAttacher cmd=/home/nino/tools/jdk-21.0.10+7/bin/java pid=235034 ino=36205999 type=java service="" logenricher=false
```
4. I checked the OBI user with:
```
ps -o user=,pid=,cmd= -C obi
```
It was still my own user.

File: pkg/internal/java/java_inject.go [hunk 1]
```diff
@@ -127,75 +131,74 @@ func (i *JavaInjector) NewExecutable(ie *ebpf.Instrumentable) error {
 	runtime.LockOSThread()
 	defer runtime.UnlockOSThread()
 
+	// We need to call cleanup here even though init may not have
+	// happened, to ensure the case when the JVM never responds and
+	// we've terminated with a timeout. Cleanup() is idempotent.
 	defer func() {
 		if err := attacher.Cleanup(); err != nil {
```
Comment (MrAlias) [inline pkg/internal/java/java_inject.go]: This cleanup can run after the in-flight helper calls `Init()` but before `Attach()` switches credentials. It clears `initialized`, so when that helper returns from `Attach()`, its deferred `Cleanup()` becomes a no-op and OBI can remain running as the target user.

Could we keep the full `Init`/`Attach`/`Cleanup` lifecycle owned by the attach goroutine, and have the timeout path wait for credentials to be restored before returning?

**Review by MrAlias — COMMENTED** [review]
The non-Java guard fixes the original trigger. The timeout path still has a credential-restoration race.

File: pkg/internal/java/java_inject.go (hunk 1 above)
Comment (grcevski) [reply pkg/internal/java/java_inject.go]: > This cleanup can run after the in-flight helper calls Init() but before Attach() switches credentials. It clears initialized, so when that helper returns from Attach(), its deferred Cleanup() becomes a no-op and OBI can remain running as the target user.

You are right, I didn't think of that one. I'm thinking maybe I don't have to flip the initialized back to false in the Cleanup() call? That way we'll restore the OBI state no matter what. 

> Could we keep the full Init/Attach/Cleanup lifecycle owned by the attach goroutine, and have the timeout path wait for credentials to be restored before returning?

Can you explain better what you mean here? I think you mean I shouldn't bother with the top level clean-up, the one I removed and then added. I was worried this Go routines stays permanently in attach somehow, flips the uid/gid to the process and something blocks it that we won't be able to cancel on. That code makes syscalls and doesn't pass the context everywhere, so if it doesn't cancel we end up with OBI with bad uid/gid.

Comment (grcevski) [reply pkg/internal/java/java_inject.go]: I think I want the following:
- The timeout ending should have a final say, it should restore the original OBI uid/gid and any changes to these shouldn't be possible after we've done that.
- I think maybe I made a mistake with moving the attach off the background. If it times out but eventually succeeds, we don't hold up the pipeline, but eventually we load the java agent or fail.
- Since I turned the actual syscalls to change uid/gid to functions, for testing reasons, I think I can easily add logic in them to say, you are changing them to the non-default, but we timed out, so back-off. I can use a lock there to prevent attach racing with the outer cleanup.

Comment (MrAlias) [reply pkg/internal/java/java_inject.go]: That makes sense, and your first bullet is exactly the invariant I was trying to describe. I was too prescriptive about waiting for the attach goroutine: if the timeout path can make credential restoration final, this concern does not require holding up the pipeline until that goroutine exits.

I don't think leaving `initialized` true by itself is enough, because restoration would still depend on the helper eventually returning. The lock approach in your latest comment sounds good if the same critical section covers the timeout check and both target credential syscalls, and the timeout path uses that lock to mark the attacher timed out and restore the original IDs. Then either attach switches first and timeout restores the IDs, or timeout wins and attach cannot switch them afterward.

In that design, I think `Cleanup()` should still clear `initialized`, so the helper's later defer becomes a no-op. I'd put the synchronization and terminal state on `JAttacher` and keep the package-level syscall variables as test seams.

Comment (grcevski) [reply pkg/internal/java/java_inject.go]: I made the changes based on my plan, except I didn't put attach in the background, I'm not sure if we should gamble with potential goroutine leak for the chance that attach succeeds later.

I introduced new field terminated and that is now what guards everything. I keep initialized as not being unset, because it's really the state of = "did we get the OBI own credentials".

File: pkg/internal/java/java_inject.go [hunk 2]
```diff
@@ -127,75 +131,74 @@ func (i *JavaInjector) NewExecutable(ie *ebpf.Instrumentable) error {
 	runtime.LockOSThread()
 	defer runtime.UnlockOSThread()
 
+	// We need to call cleanup here even though init may not have
+	// happened, to ensure the case when the JVM never responds and
+	// we've terminated with a timeout. Cleanup() is idempotent.
 	defer func() {
-		if err := attacher.Cleanup(); err != nil {
-			slog.Warn("error on JVM attach cleanup", "error", err)
+		if err := attacher.Cleanup(true); err != nil {
```
Comment (MrAlias) [inline pkg/internal/java/java_inject.go]: This timeout can still leave `NewExecutable` blocked on OpenJ9:

`JVM stops replying` → `ctx.Done()` wins → this deferred cleanup runs before the return completes → `detach()` waits without a deadline for `ATTACH_DETACHED` → the timeout error never reaches `traceAttacher`.

Credential restoration happens after `detach()`, so OBI can also remain under the target UID/GID while the discovery loop is stuck. There is a similar pre-connection gap: OpenJ9's blocking `flock(LOCK_EX)` calls do not observe `ctx` and can leave the helper plus its pinned OS thread behind after the deadline.

Could we separate graceful detach from terminal cleanup? Normal completion can send `ATTACH_DETACHED`; once the deadline fires, terminal cleanup should mark the attacher terminated, restore OBI's credentials, and interrupt its local waits without waiting for the JVM protocol. The listener, connection, and acquired locks also need to be published under the same synchronization as the terminal state, so the worker cannot acquire or publish a resource after cleanup has taken ownership. The blocking operations could then use mechanisms such as `LOCK_NB` retries that check `ctx` and socket deadlines derived from the context.

**Review by MrAlias — COMMENTED** [review]
The credential-restoration race from my prior pass is fixed. OpenJ9 can still keep the configured timeout from returning, leaving process discovery blocked.

File: pkg/internal/java/java_inject.go (hunk 2 above)
Comment (grcevski) [reply pkg/internal/java/java_inject.go]: Good point. I changed the code to separate the two paths. Essentially, everywhere where Cleanup() will be called gracefully, it's actually where we want to detach from J9. At the Terminate() new call, whcih is in the outer goroutine, I decided to not wait for the graceful detach and simply restore the credentials. This way if any of the internal goroutines finish, we'll eventually call detach and if it's not blocked, we'll clean it up, if it's blocked that goroutine will be stuck. We can work on an extension to prevent this goroutine leak by making more of the attach internals context aware, so they can cancel. But I think that belongs in another PR.

File: pkg/internal/jvmtools/jvm/cmd.go [hunk 3]
```diff
@@ -37,42 +51,91 @@ func NewJAttacher(logger *slog.Logger) *JAttacher {
...
+		return nil
+	}
+
+	var cleanupErr error
+
+	if j.j9attacher != nil {
+		cleanupErr = errors.Join(cleanupErr, j.j9attacher.detach())
+	}
+
+	cleanupErr = errors.Join(cleanupErr, j.restoreCredentialsLocked())
```
Comment (MrAlias) [inline pkg/internal/jvmtools/jvm/cmd.go]: After attach A times out, `Terminate()` attempts to restore the original IDs and `traceAttacher` continues to the next event. If A's retained worker later exits, its deferred `Cleanup()` reaches this unconditional restore again. Attach B owns a different `JAttacher`, so the per-instance mutexes do not serialize A's late restore with B's credential switches; A can reset the process-wide euid/egid while B is mid-attach.

Could we make credential ownership process-wide rather than per `JAttacher`? After a successful `Terminate()`, A's stale cleanup must not restore the IDs again. If terminal restoration fails, proceeding to B while A may retry is also unsafe, so that failure and any retry need to be coordinated with future credential switches. A shared credential lock/lease across attachers would make both cases explicit.

**Review by MrAlias — COMMENTED** [review]
NIce, I like the split between terminal cleanup and graceful OpenJ9 detach; it fixes the blocking timeout path and makes the lifecycle much easier to reason about. The PR is moving in the right direction, though one cross-attacher race remains around the process-wide credential restore.

File: pkg/internal/jvmtools/jvm/cmd.go (hunk 3 above)
Comment (grcevski) [reply pkg/internal/jvmtools/jvm/cmd.go]: Right, yeah, there's that mid attach restoration issue. I don't think I can easily make this process wide, or specific. But I have an idea. I can record on the init the process ID that currently owns the attach sequence. So if late Cleanup() comes back in to restore the credentials to OBI, it can check the owner id and if it's not them, then it will skip the restore of the credentials.

Comment (grcevski) [reply pkg/internal/jvmtools/jvm/cmd.go]: OK, I made another update and now I'm using a concept of attachID, which increments with every new attach operation. I prevent the Cleanup to restore the credentials if the active attach ID doesn't match the attachID it had during initialisation. I initially wrote the change by using the PID from FileInfo, but after some thought, I think repeated PIDs would've had the same issue. We shouldn't get the same PID to be looked at twice, but it's theoretically possible. So I think using a completely unrelated ID is better.

File: pkg/internal/jvmtools/jvm/cmd.go [hunk 4]
```diff
@@ -14,65 +14,139 @@ import (
...
+		cleanupErr = errors.Join(cleanupErr, j.j9attacher.detach())
+	}
+
+	activeAttachID := j.activeAttachID()
+
+	// Restore credentials if we are the active attach happening on the pipeline.
+	// If we were delayed and the main loop abandoned us, they would reset the
+	// credentials in Terminate and they might be in process of other JVM attach.
+	if activeAttachID == j.attachID {
+		cleanupErr = errors.Join(cleanupErr, j.restoreCredentialsLocked())
```
Comment (MrAlias) [inline pkg/internal/jvmtools/jvm/cmd.go]: `activeAttachID` is sampled before the process-wide credential restore, but nothing serializes that check with the next attach. A stale cleanup can load its own ID here, get descheduled, then attach B increments `currentAttachID` and switches to B’s UID/GID; when A resumes, it still enters this branch and restores OBI’s IDs over B’s attach. The new JVM attach can then fail under the wrong credentials, leaving that process without Java TLS telemetry even though discovery continues. The new test changes the active ID before calling the old cleanup, so it does not cover this check-then-act interleaving. Could we use a process-wide lock or lease that covers the ownership check, both restore syscalls, and the next attacher’s credential switches?

Comment (grcevski) [reply pkg/internal/jvmtools/jvm/cmd.go]: ah yeah ok, I'll change how this check is done. I think I can just change the semantics, make it verify current id and then use compare atomically.

Comment (grcevski) [reply pkg/internal/jvmtools/jvm/cmd.go]: OK, I made one more fix. Now I use a lock and I delegate the decision on whether to run the function on the main goroutine.

**Review by MrAlias — APPROVED** [review]
Awesome work 🚀
